snap/integrity/integrity.go

```diff
@@ -0,0 +1,156 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package integrity
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snap/integrity/dmverity"
+)
+
+const (
+	blockSize = 4096
+	// For now that the header only includes a fixed-size string and a fixed-size hash,
+	// the header size is always gonna be less than blockSize and will always get aligned
+	// to blockSize.
+	HeaderSize = 4096
+)
+
+var (
+	// magic is the magic prefix of snap extension blocks.
+	magic = []byte{'s', 'n', 'a', 'p', 'e', 'x', 't'}
+)
+
+// Align input `size` to closest `blockSize` value
+func align(size uint64) uint64 {
+	return (size + blockSize - 1) / blockSize * blockSize
+}
+
+// IntegrityDataHeader gets appended first at the end of a squashfs packed snap
+// before the dm-verity data. Size field includes the header size
+type IntegrityDataHeader struct {
+	Type     string        `json:"type"`
+	Size     uint64        `json:"size,string"`
+	DmVerity dmverity.Info `json:"dm-verity"`
+}
+
+// Constructs a new IntegrityDataHeader struct from a dmverity.Info struct.
+func newIntegrityDataHeader(dmVerityBlock *dmverity.Info, integrityDataSize uint64) *IntegrityDataHeader {
+	return &IntegrityDataHeader{
+		Type:     "integrity",
+		Size:     HeaderSize + integrityDataSize,
+		DmVerity: *dmVerityBlock,
+	}
+}
+
+func (integrityDataHeader IntegrityDataHeader) Encode() ([]byte, error) {
+	jsonHeader, err := json.Marshal(integrityDataHeader)
+	if err != nil {
+		return nil, err
+	}
+	logger.Debugf("integrity data header:\n%s", string(jsonHeader))
+
+	// \0 terminate
+	jsonHeader = append(jsonHeader, 0)
+
+	actualHeaderSize := align(uint64(len(magic) + len(jsonHeader) + 1))
+	if actualHeaderSize > HeaderSize {
+		return nil, fmt.Errorf("internal error: invalid integrity data header: wrong size")
+	}
+
+	header := make([]byte, HeaderSize)
+
+	copy(header, append(magic, jsonHeader...))
+
+	return header, nil
+}
+
+func (integrityDataHeader *IntegrityDataHeader) Decode(input []byte) error {
+	if !bytes.HasPrefix(input, magic) {
+		return fmt.Errorf("invalid integrity data header")
+	}
+
+	input = bytes.Trim(input, "\x00")
```

Comment:
Fixed but could you take a look if the proposed fix is good enough? I chose to ignore the -1 if a null byte is not found as the unmarshalling will fail later anyway.

Reply:
I think it would be better to return an error if no null byte is found, as although it is true that it would fail anyway, a different error message would provide more context to find the root cause of the problem.

Reply:
Changed it to return an error and also added a test for it.